packages/flutter/lib/src/material/app.dart

```diff
@@ -433,6 +434,33 @@ class MaterialApp extends StatefulWidget {
   ///  * <https://material.google.com/layout/metrics-keylines.html>
   final bool debugShowMaterialGrid;
 
+  /// A builder for inserting widgets above the [Navigator] but below the
+  /// other widgets created by the [MaterialApp] widget.
+  ///
+  /// For example, from the [BuildContext] passed to this method, the
+  /// [Directionality], [Localizations], [DefaultTextStyle], [MediaQuery], etc,
+  /// are all available. They can also be overridden in a way that impacts all
+  /// the routes in the [Navigator].
+  ///
+  /// This is rarely useful, but can be used in applications that wish to
+  /// override those defaults, e.g. to force the application into right-to-left
+  /// mode despite being in English, or to override the [MediaQuery] metrics
+  /// (e.g. to leave a gap for advertisements shown by a plugin from OEM code).
+  ///
+  /// The [builder] callback is passed two arguments, the [BuildContext] (as
+  /// `context`) and the widget representing the [Navigator] (as `child`). The
+  /// returned value should include the `child`; if it does not, then the
+  /// application will have no navigator and the [home], [routes],
+  /// [onGenerateRoute], [navigatorKey], [initialRoute], [defaultRouteName], and
+  /// [navigatorObservers] properties will have no effect.
+  ///
+  /// If [builder] is null, it is as if a builder was specified that returned
+  /// the `child` directly.
+  ///
+  /// For specifically overriding the [title] with a value based on the
+  /// [Localizations], consider [onGenerateTitle] instead.
+  final TransitionBuilder builder;
```

Comment:
Ha, I totally had this navigator problem a number of times. 

Should we call this something 'alarmingly special' like 'rootInsertBuilder' or some such to discourage use at a glance? Since this builder (that takes a context and a child) is unlike all other builders. 

Saves the user a few seconds from trying to use it and realize it's a TransitionBuilder

Reply:
AnimationBuilder.builder is a TransitionBuilder, so this is just following that pattern.

There's a number of "builder" callbacks that use different signatures, in fact.

Does that change your opinion on this?

Reply:
I guess my concern was more over the position in the hierarchy than the builder parameter type. 

Seems like other builders like LayoutBuilder, StatefulBuilder, FutureBuilder, routes etc build leafs. 

AnimatedBuilder is the only one that builds into the 'middle' of the tree so far but it's not too ambiguous because the verb is to build an animation. Arguably `builder` is the object and `child` is the proposition. 

App seems more ambiguous. `home` or `onGenerateRoute` etc are the objects. *App.builder sounds too invitingly mainstream wrt what it's really for (unless perhaps if we removed the `assert(onGenerateRoute != null)` if you intended this to be a mainstream alternative to using the internal Navigator building mechanism).

Reply:
I'll submit a patch that does the copyWith thing instead.

Reply:
They build leaves in that they're not passed a widget that is expected to be inserted, but other than that, they're not that different...

Anyway, I guess we could call it something else, but what? `rootInsertBuilder` is pretty ugly. :-)